src/codegen/datatype/datatype_registry.cc

```diff
@@ -0,0 +1,123 @@
+    ss << DatatypeRegistry::Global()->GetTypeName(type_code);
+  } else {
+    ss << runtime::TypeCode2Str(type_code);
+  }
+
+  ss << ".";
+
+  if (DatatypeRegistry::Global()->DatatypeRegistered(src_type_code)) {
+    ss << DatatypeRegistry::Global()->GetTypeName(src_type_code);
+  } else {
+    ss << runtime::TypeCode2Str(src_type_code);
+  }
+
+  return runtime::Registry::Get(ss.str());
+}
+
+TVM_REGISTER_GLOBAL("_register_Cast")
+    .set_body([](TVMArgs args, TVMRetValue *rv) {
+      const std::string target = args[0];
+      const std::string type = args[1];
+      const std::string src_type = args[2];
+      const std::string extern_func_name = args[3];
+
+      auto lower_cast_name =
+          "tvm.datatypes.lower." + target + ".Cast." + type + "." + src_type;
+      runtime::Registry::Register(lower_cast_name)
+          .set_body([extern_func_name](TVMArgs args, TVMRetValue *rv) {
+            Expr e = args[0];
+            const ir::Cast *cast = e.as<ir::Cast>();
+            internal_assert(cast);
+            // Custom datatypes should get cast to their underlying storage
+            // type.
+            // TODO(gus) I'm not using the width registered originally; I'm
+            // using the bits() attached to the type (where does this come
+            // from?)
+            auto return_type = DatatypeRegistry::Global()->DatatypeRegistered(
+                                   cast->type.code())
+                                   ? UInt(cast->type.bits())
+                                   : cast->type;
+            *rv = ir::Call::make(return_type, extern_func_name, {cast->value},
+                                 ir::Call::Extern);
+          });
+    });
+
+#define REGISTER_OP_A_B(OP)                                                    \
+  TVM_REGISTER_GLOBAL("_register_" #OP)                                        \
+      .set_body([](TVMArgs args, TVMRetValue *rv) {                            \
+        const std::string target = args[0];                                    \
+        const std::string type = args[1];                                      \
+        const std::string extern_func_name = args[2];                          \
+        auto lower_op_name =                                                   \
+            "tvm.datatypes.lower." + target + "." #OP "." + type;              \
+        runtime::Registry::Register(lower_op_name)                             \
+            .set_body([extern_func_name](TVMArgs args, TVMRetValue *rv) {      \
+              Expr e = args[0];                                                \
+              const ir::OP *op = e.as<ir::OP>();                               \
+              internal_assert(op);                                             \
+              *rv = ir::Call::make(UInt(op->type.bits()), extern_func_name,    \
+                                   {op->a, op->b}, ir::Call::Extern);          \
```

Comment:
we can do this all in Python -- we can 

tvm.dtype.register_lowering_rule
look at register intrinsic rule

Reply:
Basically, push as much of the code into Python as possible.
This above function can be replicated in Python instead of C++. 
LowerDatatypes will still be just about exactly the same.
I think the Python code in my sample repo will also not change all that much.